src/coreclr/src/vm/object.cpp

```diff
@@ -335,62 +335,43 @@ void STDCALL CopyValueClassUnchecked(void* dest, void* src, MethodTable *pMT)
 
     _ASSERTE(!pMT->IsArray());  // bunch of assumptions about arrays wrong.
 
-    // <TODO> @todo Only call MemoryBarrier() if needed.
-    // Reflection is a known use case where this is required.
-    // Unboxing is a use case where this should not be required.
-    // </TODO>
-    MemoryBarrier();
-
-        // Copy the bulk of the data, and any non-GC refs.
-    switch (pMT->GetNumInstanceFieldBytes())
+    if (pMT->ContainsPointers())
     {
-    case 1:
-        *(UINT8*)dest = *(UINT8*)src;
-        break;
-#ifndef ALIGN_ACCESS
-        // we can hit an alignment fault if the value type has multiple
-        // smaller fields.  Example: if there are two I4 fields, the
-        // value class can be aligned to 4-byte boundaries, yet the
-        // NumInstanceFieldBytes is 8
-    case 2:
-        *(UINT16*)dest = *(UINT16*)src;
-        break;
-    case 4:
-        *(UINT32*)dest = *(UINT32*)src;
-        break;
-    case 8:
-        *(UINT64*)dest = *(UINT64*)src;
-        break;
-#endif // !ALIGN_ACCESS
-    default:
-        memcpyNoGCRefs(dest, src, pMT->GetNumInstanceFieldBytes());
-        break;
+        memmoveGCRefs(dest, src, pMT->GetNumInstanceFieldBytes());
     }
-
-        // Tell the GC about any copies.
-    if (pMT->ContainsPointers())
+    else
     {
-        CGCDesc* map = CGCDesc::GetCGCDescFromMT(pMT);
-        CGCDescSeries* cur = map->GetHighestSeries();
-        CGCDescSeries* last = map->GetLowestSeries();
-        DWORD size = pMT->GetBaseSize();
-        _ASSERTE(cur >= last);
-        do
+        // <TODO> @todo Only call MemoryBarrier() if needed.
+        // Reflection is a known use case where this is required.
+        // Unboxing is a use case where this should not be required.
+        // </TODO>
+        MemoryBarrier();
```

Comment:
Should this be on the ContainsPointers path as well?

Reply:
ContainsPointers has a similar barrier before it does actual copying. 

I do not know why this is here though. A full fence is typically paired to a similar one which may run on a different thread, or something to a similar effect. The comment does not where that is, just says it is needed, so I kept it. 

The barrier in ContainsPointers path is a bit strange too. - If we would be copying element-wise in a for loop, we would not be doing full fence, so I am not sure.